### PR TITLE
fix: reset APT/LRPT decoder state between auto-record passes (closes #544)

### LINKED ISSUE #544: Reset LRPT/APT decoder state between auto-record passes (not just on source stop)
## Context

The DSP-side decoder state for both APT and LRPT auto-record currently only gets reset on full source stop (the `cleanup` function in `crates/sdr-core/src/controller.rs`). Between back-to-back passes — e.g., a user listening to FM with auto-record on, three Meteor passes arriving over the next 6 hours — the decoder + pipeline + assembler retain state across pass boundaries.

For LRPT specifically:
- `LrptPipeline`'s internal `ImageAssembler` keeps every previous pass's per-channel pixels in memory (~6 MB per channel per pass).
- `LrptDecoder::last_pushed_lines` watermark grows monotonically.
- Drawing still works correctly by coincidence: the pipeline's cumulative line counter and the decoder's watermark stay in lockstep, so the diff is always \"just the new lines\" of the current pass. The shared `LrptImage` gets cleared at AOS by the wiring layer (`state.lrpt_image.clear()`), so the visible canvas is fresh.
- But the pipeline's internal assembler is a distinct object from the shared image — it's never cleared between passes. Memory grows linearly with pass count.

For APT this is mostly invisible because `RestoreTune` often closes the source (when `was_running == false` pre-AOS), which triggers `cleanup` and resets the APT decoder. But if the user was already running pre-AOS, the same cumulative-state issue exists.

## Proposed fix

Add a between-pass reset hook that the wiring layer calls at LOS. Two surfaces:

1. **New `UiToDsp` variant** — e.g. `UiToDsp::ResetImagingDecoders` — that flushes both `state.apt_decoder` and `state.lrpt_decoder`. Sent at LOS alongside the existing `RestoreTune` action.
2. **Recorder action** — extend `Action` with `Action::ResetImagingDecoders` so the state machine emits it at the `Recording → Finalizing` transition. Wiring layer maps it to the new `UiToDsp` message. Symmetric with `StopAutoAudioRecord` which is also a between-pass cleanup.

## Touchpoints

- `crates/sdr-core/src/messages.rs` — add `ResetImagingDecoders` (or similar) to `UiToDsp`.
- `crates/sdr-core/src/controller.rs` — handle the new message; for LRPT, call `decoder.reset()` and let it propagate via the existing `Err → drop for re-init` path; for APT, call `decoder.reset()` directly. Don't drop the decoders entirely — that wastes the filter taps + resampler tables (same rationale as the existing source-stop reset).
- `crates/sdr-ui/src/sidebar/satellites_recorder.rs` — emit the new `Action` variant from `tick_recording`/`tick_before_pass` LOS branches.
- `crates/sdr-ui/src/window.rs` — handle the new `RecorderAction` variant in the dispatcher.

## Tests

- Recorder unit test: pin that LOS emits the reset action.
- Integration: simulate two back-to-back passes (drive the recorder through AOS₁ → LOS₁ → AOS₂ → LOS₂) and assert two reset actions fire.

## Priority

Low. Bounded by SDR session lifetime; most users won't notice. File before users with overnight unattended setups hit it.

Surfaced in PR #543 review.

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

The auto-record flow now flushes APT + LRPT decoder state at every LOS, not just on full source stop. Closes #544.

## Why

Pre-this-PR, decoder reset only happened in `cleanup` (full source stop). That's fine when `was_running == false` pre-AOS — `RestoreTune` at LOS calls `set_playing(false)` and the source-stop path runs. But when the user was already running pre-AOS (e.g. listening to FM with auto-record on, three Meteor passes overnight), the source stays open across LOS → AOS, `cleanup` never fires, and:

- `LrptPipeline`'s internal `ImageAssembler` retains every previous pass's pixels (~6 MB per channel per pass).
- `LrptDecoder::last_pushed_lines` watermark grows monotonically.
- `AptDecoder` accumulator + ready queue keep growing.

Memory grows linearly with pass count. Surfaced in PR #543 review.

## Approach

1. Lift the existing APT + LRPT reset block out of `cleanup` into a `reset_imaging_decoders(state)` helper. Single source of truth — `cleanup` calls it for source-stop, the new handler calls it for between-pass.
2. Add `UiToDsp::ResetImagingDecoders`. Handler is a one-liner over the helper.
3. Add `Action::ResetImagingDecoders` to the recorder's action enum. Emit it at every `Recording → Finalizing` transition (LOS), AFTER the save action so the PNG export still snapshots the just-finished pass.
4. Wire the new action variant in `connect_satellites_panel`'s `interpret_action`.

## Idempotency

When `was_running == false` pre-AOS, the LOS chain fires the reset, then `RestoreTune` calls `set_playing(false)` which triggers the source-stop path which also calls `reset_imaging_decoders`. The second call is a no-op (`AptDecoder::reset` clears already-empty buffers, `LrptDecoder::reset` is similarly stateless on already-flushed input) — the double-reset is harmless.

## Commits

1. `sdr-core: extract reset_imaging_decoders helper + UiToDsp::ResetImagingDecoders` — 7ee47ed. Refactor + new variant + handler + shape-regression test.
2. `sdr-ui: emit ResetImagingDecoders at LOS + wiring + tests` — 308e2d7. Action variant + state-machine emit + window.rs wiring + two new pure-state-machine tests.

## Test plan

- [x] `cargo build --workspace --features whisper-cpu`
- [x] `cargo clippy --workspace --all-targets --features whisper-cpu -- -D warnings`
- [x] `cargo test --workspace --features whisper-cpu` — all 31 recorder tests pass including the two new ones (`los_emits_reset_imaging_decoders`, `two_back_to_back_passes_each_emit_reset`)
- [x] `cargo check --workspace --no-default-features --features sherpa-cpu` (mutex cross-check)
- [x] `cargo fmt --all -- --check`
- [ ] Manual smoke (deferred to next pass): APT/LRPT pass at LOS — confirm new tracing line `auto-record LOS: resetting imaging decoders` appears; back-to-back passes (start source pre-AOS so `was_running == true`) produce two clean PNGs without pass-1 pixels bleeding into pass-2.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Auto-record now issues an imaging-decoder reset between passes at loss-of-signal, clearing APT and LRPT decoder buffers while keeping the receiver active for the next pass.
* **Bug Fixes**
  * Unified between-pass and full-stop reset behavior to prevent stale decoder state and ensure consistent cleanup; retains warning-and-drop if LRPT reset fails.
* **Tests**
  * Added regression tests to verify reset is emitted once per pass, preserves save-before-reset ordering, and covers edge-case pass transitions.

### CODE REVIEW

File: crates/sdr-ui/src/sidebar/satellites_recorder.rs [hunk 1]
```diff
@@ -609,11 +620,24 @@ impl AutoRecorder {
...
+            // the subsequent `set_playing(false)` in
+            // `RestoreTune` will trigger the source-stop path
+            // which resets again — idempotent, so the
+            // double-reset is fine. Per issue #544.
+            let mut actions = Vec::with_capacity(3);
             actions.push(save_action_for(&output));
             if audio_path.is_some() {
                 actions.push(Action::StopAutoAudioRecord);
             }
+            actions.push(Action::ResetImagingDecoders);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/satellites_recorder.rs]: Done in b46e31d. Extracted `los_actions_for(output, has_audio)` as the single source of truth for the LOS-finalisation action vec. Both `tick_recording`'s end-of-pass LOS branch AND `tick_before_pass`'s stall short-circuit now go through it, so the BeforePass-stall edge case (1 Hz driver stalled, or pass entirely inside the settle window) gets the same `ResetImagingDecoders` flush. Helper documents the ordering (save → stop audio → reset decoders) and the idempotency contract with the source-stop reset, so future readers can't reintroduce the drift. Regression test `los_during_before_pass_still_emits_reset_imaging_decoders` exercises the stalled-BeforePass path explicitly — 60 s pass, jump past LOS without ticking through settle, assert the action vec contains `Action::ResetImagingDecoders`. 32 recorder tests pass.

File: crates/sdr-ui/src/sidebar/satellites_recorder.rs [hunk 2]
```diff
@@ -276,6 +276,17 @@ pub enum Action {
     /// pass does NOT retroactively start or stop recording.
     /// Wiring layer maps to `UiToDsp::StopAudioRecording`.
     StopAutoAudioRecord,
+    /// Flush in-flight imaging-decoder state between passes.
+    /// Fired alongside [`Action::SavePng`] /
+    /// [`Action::SaveLrptPass`] on the `Recording → Finalizing`
+    /// transition. Without it, when the user was already
+    /// running pre-AOS (`SavedTune.was_running == true`), the
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/satellites_recorder.rs]: Done in b785242. Updated the `ResetImagingDecoders` doc comment to mention both LOS transition paths (`Recording → Finalizing` end-of-pass + `BeforePass → Finalizing` stall short-circuit), referenced `los_actions_for` as the shared builder, and noted the save → optional-stop-audio → reset ordering it guarantees. Now reads in lockstep with the actual emit sites.

File: crates/sdr-ui/src/sidebar/satellites_recorder.rs [hunk 3]
```diff
@@ -1786,4 +1868,115 @@ mod tests {
...
+            default_tune(),
+        );
+        let los_plus = pass.end + ChronoDuration::seconds(1);
+        let los_actions = r.tick(los_plus, &[pass], true, false, default_tune());
+        assert!(
+            los_actions
+                .iter()
+                .any(|a| matches!(a, Action::ResetImagingDecoders)),
+            "LOS must emit Action::ResetImagingDecoders; got {los_actions:?}",
+        );
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/satellites_recorder.rs]: Done in b785242. Added `assert_save_before_reset(actions, label)` test helper that pins the LOS contract: save action runs BEFORE reset action (otherwise the save's snapshot read of the shared `LrptImage` would capture an empty buffer instead of the just-finished pass's pixels). Helper applied at all four LOS-reset assertion sites — `los_emits_reset_imaging_decoders`, both pass legs of `two_back_to_back_passes_each_emit_reset`, and `los_during_before_pass_still_emits_reset_imaging_decoders` — so the BeforePass-stall path also gets ordering coverage. Single helper instead of duplicating the position-based check four times; failure messages carry a label identifying which LOS path tripped.
